fccloudapi.php

```diff
@@ -171,56 +171,92 @@ public function getObjectName()
-    const TYPE_BOOLEAN = 4;
-    const TYPE_DATE = 5;
-    const TYPE_ENUMERATION = 6;
-    const TYPE_ARRAY = 7;
-
     /**
+     * Cast response data to type
+     *
      * @param mixed $data
      * @param int $type
      * @return array|bool|\DateTime|float|int
      */
-    protected function castToType($data, int $type)
+    public function castToType($data, int $type)
     {
         switch ($type) {
-            case self::TYPE_INTEGER:
-                return (int) $data;
-            case self::TYPE_DECIMAL:
-                return (float) $data;
-            case self::TYPE_BOOLEAN:
+            case AbstractMetadataRecord::TYPE_INTEGER:
+                return strlen($data) ? (int) $data : null;
+            case AbstractMetadataRecord::TYPE_DECIMAL:
+                return strlen($data) ? (float) $data : null;
+            case AbstractMetadataRecord::TYPE_BOOLEAN:
                 return !!json_decode($data);
-            case self::TYPE_DATE:
-                return \DateTime::createFromFormat('Y-m-d H:i:s', $data);
-            case self::TYPE_ARRAY:
-                return explode(',', $data);
+            case AbstractMetadataRecord::TYPE_DATE:
+                $date = \DateTime::createFromFormat('Y-m-d H:i:s', $data);
+                
+                return $date instanceof \DateTime ? $date : null;
+            case AbstractMetadataRecord::TYPE_ARRAY:
+                return strlen($data) ? explode(',', $data) : [];
             default:
                 return $data;
         }
     }
+
+    /**
+     * Best effort reverse cast of attribute value for submission
+     * to api endpoint.
+     *
+     * @param $data
+     * @param int $type
+     * @return string|int|float|bool
+     */
+    protected function reverseCastFromType($data, int $type)
+    {
+        switch ($type) {
+            case AbstractMetadataRecord::TYPE_DATE:
+                return $data instanceof \DateTime ? $data->format('Y-m-d H:i:s') : $data;
+            case AbstractMetadataRecord::TYPE_INTEGER:
+            case AbstractMetadataRecord::TYPE_DECIMAL:
+            case AbstractMetadataRecord::TYPE_BOOLEAN:
+                return json_encode($data);
+            case AbstractMetadataRecord::TYPE_ARRAY:
+                return is_array($data) ? implode(',', $data): $data;
+        }
+
+        return $data;
+    }
 }
 
 /**
- * Class MetadataSetRecord
+ * Hold common constants, methods, etc.
+ * @package codelathe\fccloudapi
+ */
+abstract class AbstractMetadataRecord extends DataRecord
+{
+    const TYPE_TEXT = 1;
+    const TYPE_INTEGER = 2;
+    const TYPE_DECIMAL = 3;
+    const TYPE_BOOLEAN = 4;
+    const TYPE_DATE = 5;
+    const TYPE_ENUMERATION = 6;
+    const TYPE_ARRAY = 7;
+
+    use MetadataAttributeTypeCasterTrait;
+}
+
+/**
+ * Hold common constants, methods, etc.
  * @package codelathe\fccloudapi
  */
-final class MetadataSetRecord extends AbstractMetadataRecord
+abstract class BaseMetadataSetRecord extends AbstractMetadataRecord
```

Comment:
I'm not a huge fan of that abstraction as well. IMO it serves no purpose other than extracting some properties that are shared between different object, which in fact have nothing in common. This something that maybe narrows the amount of code but also makes it a lot harder to read. I like the usage of trait, which I think is great for that scenario.

Reply:
Sure, I think I can see it that way also. It was quite easy to fall into the assumption they were related due to the handful of fields they shared. Are you suggesting we share common properties & methods via a trait instead? I haven't originally thought of doing so but it does reduce the depth of inheritance which is good.